### PR TITLE
Rule S5689 Recovering fingerprints from web application technologies should not be possible

### LINKED ISSUE #1836: Rule S5689 Recovering fingerprints from web application technologies should not be possible
Implements [RSPEC-5689](https://jira.sonarsource.com/browse/RSPEC-5689)

For both JS/TS

Code examples: https://github.com/SonarSource/security-expected-issues/tree/master/javascript/rules/vulnerabilities/RSPEC-5689

Express.js app is considered safe if any of the following is executed:
```
app.disable("x-powered-by");
```
or 
```
let hidePoweredBy = require('hide-powered-by'); 
app.use(hidePoweredBy());
```

Report on the `app.express()` instantiation if both are missing

### PR DESCRIPTION
**Pull request by andrey-tyukin-sonarsource**
Fixes #1836

### CODE REVIEW

File: eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+import { RuleTester } from 'eslint';
+
+const ruleTester = new RuleTester({
+  parser: require.resolve('@typescript-eslint/parser'),
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts:23]: why not using default parser?

File: eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+  parserOptions: { ecmaVersion: 2018, sourceType: 'module' },
+});
+
+import { rule } from '../../src/rules/express-header-hide-x-powered-by';
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts:27]: remove `../../src/`

File: eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+ruleTester.run('X-Powered-By should be hidden', rule, {
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts:29]: we usually put rule title here

File: eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+  valid: [
+    {
+      code: `
+        var express = require('express')
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts:33]: I think we should prefer `let`/`const` declaration in tests (as it is more common in new js code) and putting `;` after each statement

File: eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        var app = express()
+        app.set("x-powered-by", true) // Noncompliant
+      `,
+      errors: 1,
+    },
+    {
+      code: `
+        var express = require('express')
+        var app = express()
+        var coverage = true, only = true;
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts:133]: what are we testing here?

File: eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+        var express = require('express')
+        var app = express()
+        app.set("X-PoWeReD-bY", false) // Compliant
+      `,
+    },
+    {
+      code: `
+        var express = require('express')
+        var app = express()
+        app.use(require('helmet')()) // Compliant
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts:64]: let's add test where middleware is instantiated but not passed in `app.use`

File: eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    {
+      code: `
+        import express from 'express';
+        import disableXPoweredBy from 'hide-powered-by';
+        const app = express();
+        app.use(disableXPoweredBy()) // Compliant
+      `,
+    },
+  ],
+  invalid: [
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts:76]: let's add test with 2 apps (one safe and one unsafe) to show this limitation

File: eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+        const hidePoweredBy = require('hide-powered-by')
+        var app = express()
+        app.use(hidePoweredBy()) // Compliant
+      `,
+    },
+    {
+      code: `
+        var express = require('express')
+        var app = express()
+        app.set("x-powered-by", false) // Compliant
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts:50]: did you discuss with @eric-therond-sonarsource if we should mention this patten and pattern with `helmet` in the rule description?

File: eslint-bridge/src/rules/express-header-hide-x-powered-by.ts [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+
+/**
+ * Checks whether the expression looks somewhat like `app.use(m())`,
+ * where `m` is from the list of allowed middleware.
+ */
+function checkUseMiddleware(
+  callExpression: estree.CallExpression,
+  context: Rule.RuleContext,
+  app: estree.Identifier,
+  middleware: string[],
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/express-header-hide-x-powered-by.ts:99]: why we need this parameter if it's always the same value?

File: eslint-bridge/src/rules/express-header-hide-x-powered-by.ts [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+// https://jira.sonarsource.com/browse/RSPEC-5689
+
+import { Rule } from 'eslint';
+import * as estree from 'estree';
+import { getModuleNameOfNode } from './utils';
+
+// Limitations:
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/express-header-hide-x-powered-by.ts:26]: add tests to show these limitations (and I would drop these comments then)

File: eslint-bridge/src/rules/express-header-hide-x-powered-by.ts [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+    }
+  }
+  return undefined;
+}
+
+/**
+ * Checks whether the expression looks somewhat like `app.use(m())`,
+ * where `m` is from the list of allowed middleware.
+ */
+function checkUseMiddleware(
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/express-header-hide-x-powered-by.ts:95]: rename these boolean-returning functions to start with `is`

File: eslint-bridge/src/rules/express-header-hide-x-powered-by.ts [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+export const rule: Rule.RuleModule = {
+  create(context: Rule.RuleContext) {
+    let appInstantiation: estree.Identifier | null = null;
+    let isSafe = false;
+    return {
+      CallExpression: (node: estree.Node) => {
+        if (!isSafe && appInstantiation) {
+          const callExpr = node as estree.CallExpression;
+          isSafe =
+            checkUseMiddleware(callExpr, context, appInstantiation, PROTECTING_MIDDLEWARE) ||
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/express-header-hide-x-powered-by.ts:46]: we discussed with Tibor that we want `context` to be first parameter in such functions

File: eslint-bridge/src/rules/express-header-hide-x-powered-by.ts [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+      const usedMiddleware = String(getModuleNameOfNode(arg.callee, context)?.value);
+      if (usedMiddleware) {
+        return middleware.includes(usedMiddleware);
+      }
+    }
+  }
+  return false;
+}
+
+function checkDisable(callExpression: estree.CallExpression, app: estree.Identifier): boolean {
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/express-header-hide-x-powered-by.ts:113]: tbh the name is not very helpful, can we say what is disabled?

File: eslint-bridge/src/rules/express-header-hide-x-powered-by.ts [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+  if (isMethodInvocation(callExpression, app.name, 'disable', 1)) {
+    const arg0 = callExpression.arguments[0];
+    return arg0.type === 'Literal' && String(arg0.value).toLowerCase() === HEADER_X_POWERED_BY;
+  }
+  return false;
+}
+
+function checkSetFalse(callExpression: estree.CallExpression, app: estree.Identifier): boolean {
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/express-header-hide-x-powered-by.ts:121]: again about name, change it so that it's clear what is set to false

File: eslint-bridge/src/rules/express-header-hide-x-powered-by.ts [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+        if (!isSafe && appInstantiation) {
+          const callExpr = node as estree.CallExpression;
+          isSafe =
+            checkUseMiddleware(callExpr, context, appInstantiation, PROTECTING_MIDDLEWARE) ||
+            checkDisable(callExpr, appInstantiation) ||
+            checkSetFalse(callExpr, appInstantiation);
+        }
+      },
+      VariableDeclaration: (node: estree.Node) => {
+        if (!isSafe && appInstantiation == null) {
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/express-header-hide-x-powered-by.ts:52]: use `!appInstantiation` to be consistent

File: eslint-bridge/src/rules/express-header-hide-x-powered-by.ts [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+      },
+    };
+  },
+};
+
+function attemptFindExpressAppInstantiation(
+  varDecl: estree.VariableDeclaration,
+  context: Rule.RuleContext,
+): estree.Identifier | undefined {
+  if (varDecl.declarations.length === 1) {
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/express-header-hide-x-powered-by.ts:76]: I don't get why are we forcing this condition? what's wrong with `let app = express(), zoo = 42`?
I think we should support it as well. And I would subscribe to `VariableDeclarator` instead to ease the logic

File: eslint-bridge/src/rules/express-header-hide-x-powered-by.ts [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+    const onlyDecl = varDecl.declarations[0];
+    const rhs = onlyDecl.init;
+    if (rhs && rhs.type === 'CallExpression') {
+      const rhsCall = rhs as estree.CallExpression;
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/express-header-hide-x-powered-by.ts:80]: can be inlined

File: eslint-bridge/src/rules/express-header-hide-x-powered-by.ts [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+      const callee = rhsCall.callee;
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/express-header-hide-x-powered-by.ts:81]: destructuring can be used: `const { callee } = rhsCall`

File: eslint-bridge/src/rules/express-header-hide-x-powered-by.ts [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+      if (getModuleNameOfNode(callee, context)?.value === 'express') {
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/express-header-hide-x-powered-by.ts:82]: how about creating constant for `express`?

File: eslint-bridge/src/rules/express-header-hide-x-powered-by.ts (hunk 11 above)
Comment (vilchik-elena) [inline eslint-bridge/src/rules/express-header-hide-x-powered-by.ts:95]: I think with this implementation we don't support following pattern
```
const helmet = require('helmet');
const helmet1 = helmet();
app.use(helmet1);
```

File: eslint-bridge/src/rules/express-header-hide-x-powered-by.ts [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+import * as estree from 'estree';
+import { getModuleNameOfNode } from './utils';
+
+// Limitations:
+// 1. `app.use(helmet.hidePoweredBy());` is not recognized -> FP danger
+// 2. Options passed to `helmet` are not checked: maybe `hidePoweredBy` is switched off -> FN danger
+
+const HELMET = 'helmet';
+const HIDE_POWERED_BY = 'hide-powered-by';
+const HEADER_X_POWERED_BY = 'X-Powered-By'.toLowerCase();
```
[the remark this answers is not in the document]
Comment (andrey-tyukin-sonarsource) [reply eslint-bridge/src/rules/express-header-hide-x-powered-by.ts:32]: HTTP Headers are case insensitive: [question on SO](https://stackoverflow.com/questions/5258977/are-http-headers-case-sensitive), i.e. it doesn't matter whether the user invokes `app.disable('X-Powered-By')` or `app.disable('x-powered-by')`. I've seen the lowercase version multiple times.

File: javascript-checks/src/main/java/org/sonar/javascript/checks/HidePoweredByCheck.java [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+package org.sonar.javascript.checks;
+
+import org.sonar.check.Rule;
+import org.sonar.javascript.checks.annotations.JavaScriptRule;
+import org.sonar.javascript.checks.annotations.TypeScriptRule;
+
+@JavaScriptRule
+@TypeScriptRule
+@Rule(key = "S5689")
+public class HidePoweredByCheck extends EslintBasedCheck {
```
[the remark this answers is not in the document]
Comment (andrey-tyukin-sonarsource) [reply javascript-checks/src/main/java/org/sonar/javascript/checks/HidePoweredByCheck.java:29]: Since the entire check is concerned with X-Powered-By header, I've renamed it to XPoweredByCheck, that seems the least unexpected variant.

File: eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts (hunk 8 above)
Comment (andrey-tyukin-sonarsource) [reply eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts]: Regarding the improvements that were implemented, I got the response ["nice improvements"](https://github.com/SonarSource/SonarJS/issues/1836#issuecomment-674155003).

File: eslint-bridge/src/rules/express-header-hide-x-powered-by.ts (hunk 9 above)
Comment (andrey-tyukin-sonarsource) [reply eslint-bridge/src/rules/express-header-hide-x-powered-by.ts]: Because the method body does not depend in any way on a particular choice of middlewares, so this is reflected in the signature. It also makes the method reusable: as soon as I encounter a similar case again, I'll consider reusing it.

File: eslint-bridge/src/rules/express-header-hide-x-powered-by.ts (hunk 10 above)
Comment (andrey-tyukin-sonarsource) [reply eslint-bridge/src/rules/express-header-hide-x-powered-by.ts]: The second one is actually not a problem for this particular header: "This middleware takes no options.". Added test for first one.

**Comment by andrey-tyukin-sonarsource** [comment]
Checking whether `app` escapes into some method (4b43123) was discussed with Eric Therond.

File: eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts (hunk 7 above)
Comment (vilchik-elena) [reply eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts]: I don't find such test

File: eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts (hunk 8 above)
Comment (vilchik-elena) [reply eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts]: I don't question your improvements, I've asked if rule description should be updated. Eric added `helmet` to it, please update

File: eslint-bridge/src/rules/express-header-hide-x-powered-by.ts (hunk 11 above)
Comment (vilchik-elena) [reply eslint-bridge/src/rules/express-header-hide-x-powered-by.ts]: I was implying that we should update implementation to support it :)

File: eslint-bridge/src/rules/express-header-hide-x-powered-by.ts [hunk 22]
```diff
@@ -0,0 +1,148 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+// https://jira.sonarsource.com/browse/RSPEC-5689
+
```
[the remark this answers is not in the document]
Comment (andrey-tyukin-sonarsource) [reply eslint-bridge/src/rules/express-header-hide-x-powered-by.ts]: Renamed all `x-powered-by`-related things to `x-powered-by` and `XPoweredBy`.

File: eslint-bridge/src/rules/express-header-hide-x-powered-by.ts [hunk 23]
```diff
@@ -0,0 +1,148 @@
...
+// Limitations:
+// 1. `app.use(helmet.hidePoweredBy());` is not recognized -> FP danger
+// 2. Options passed to `helmet` are not checked: maybe `hidePoweredBy` is switched off -> FN danger
+
+const HELMET = 'helmet';
+const HIDE_POWERED_BY = 'hide-powered-by';
+const HEADER_X_POWERED_BY = 'X-Powered-By'.toLowerCase();
+const MESSAGE = 'Disable the fingerprinting of this web technology.';
+const PROTECTING_MIDDLEWARE = [HELMET, HIDE_POWERED_BY];
+const NUM_ARGS_SET = 2;
```
[the remark this answers is not in the document]
Comment (andrey-tyukin-sonarsource) [reply eslint-bridge/src/rules/express-header-hide-x-powered-by.ts]: Done.

File: eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts (hunk 1 above)
Comment (andrey-tyukin-sonarsource) [reply eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts]: Copy-paste artifact. Removed.

File: eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts (hunk 3 above)
Comment (andrey-tyukin-sonarsource) [reply eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts]: Copied title from ticket description.

File: eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts (hunk 4 above)
Comment (andrey-tyukin-sonarsource) [reply eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts]: replaced everything by consts

File: eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts (hunk 5 above)
Comment (andrey-tyukin-sonarsource) [reply eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts]: Code coverage corner case. No longer needed.

File: eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts (hunk 6 above)
Comment (andrey-tyukin-sonarsource) [reply eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts]: This is the test with comment `// imported, but forgot to use helmet`.

File: eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts (hunk 7 above)
Comment (andrey-tyukin-sonarsource) [reply eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts]: I don't understand what this comment is referring to. If it's referring to previous comment: there is an invalid test where `helmet` is instantiated but left unused.

File: eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts (hunk 8 above)
Comment (andrey-tyukin-sonarsource) [reply eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts]: Oh, he's also added `helmet.hidePoweredBy()`. This will have to be added, after all.

File: eslint-bridge/src/rules/express-header-hide-x-powered-by.ts (hunk 11 above)
Comment (andrey-tyukin-sonarsource) [reply eslint-bridge/src/rules/express-header-hide-x-powered-by.ts]: `isUsingMiddleware` now.

File: eslint-bridge/src/rules/express-header-hide-x-powered-by.ts (hunk 12 above)
Comment (andrey-tyukin-sonarsource) [reply eslint-bridge/src/rules/express-header-hide-x-powered-by.ts]: moved.

File: eslint-bridge/src/rules/express-header-hide-x-powered-by.ts (hunk 13 above)
Comment (andrey-tyukin-sonarsource) [reply eslint-bridge/src/rules/express-header-hide-x-powered-by.ts]: `isDisabledXPoweredBy` + `isSetFalseXPoweredBy` now.

File: eslint-bridge/src/rules/express-header-hide-x-powered-by.ts (hunk 14 above)
Comment (andrey-tyukin-sonarsource) [reply eslint-bridge/src/rules/express-header-hide-x-powered-by.ts]: It's `isSetFalseXPoweredBy` now.

File: eslint-bridge/src/rules/express-header-hide-x-powered-by.ts (hunk 15 above)
Comment (andrey-tyukin-sonarsource) [reply eslint-bridge/src/rules/express-header-hide-x-powered-by.ts]: ok: `if (!isSafe && !appInstantiation)`

File: eslint-bridge/src/rules/express-header-hide-x-powered-by.ts (hunk 16 above)
Comment (andrey-tyukin-sonarsource) [reply eslint-bridge/src/rules/express-header-hide-x-powered-by.ts]: Changed to VariableDeclarator, that was a really good idea. Also, it removed some strange corner cases that were unreachable by coverage.

File: eslint-bridge/src/rules/express-header-hide-x-powered-by.ts (hunk 17 above)
Comment (andrey-tyukin-sonarsource) [reply eslint-bridge/src/rules/express-header-hide-x-powered-by.ts]: `const { callee } = rhs as estree.CallExpression;`

File: eslint-bridge/src/rules/express-header-hide-x-powered-by.ts (hunk 18 above)
Comment (andrey-tyukin-sonarsource) [reply eslint-bridge/src/rules/express-header-hide-x-powered-by.ts]: `const { callee } = rhs as estree.CallExpression;`

File: eslint-bridge/src/rules/express-header-hide-x-powered-by.ts (hunk 19 above)
Comment (andrey-tyukin-sonarsource) [reply eslint-bridge/src/rules/express-header-hide-x-powered-by.ts]: Done.

File: eslint-bridge/src/rules/express-header-hide-x-powered-by.ts (hunk 11 above)
Comment (andrey-tyukin-sonarsource) [reply eslint-bridge/src/rules/express-header-hide-x-powered-by.ts]:
```
{
        code: `
        const express = require('express');
        const helmet = require('helmet');
        const helmetInst = helmet();
        const app = express();
        app.use(helmetInst); // compliant
      `,
      },
```
It's now supported. Here is the test, with `helmetInst` instead of `helmet1`.

File: eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts (hunk 8 above)
Comment (andrey-tyukin-sonarsource) [reply eslint-bridge/tests/rules/express-header-hide-x-powered-by.test.ts]: Updated metadata, implemented `helmet.hidePoweredBy()`-case.

File: eslint-bridge/src/rules/x-powered-by.ts [hunk 24]
```diff
@@ -0,0 +1,234 @@
...
+      return flatMap(noDerefNode.elements, recHelper);
+    } else {
+      return [noDerefNode];
+    }
+  }
+
+  return flatMap(appUseArguments, recHelper);
+}
+
+function flatMap<A, B>(xs: A[], f: (e: A) => B[]): B[] {
```
Comment (saberduck) [inline eslint-bridge/src/rules/x-powered-by.ts:139]: I was really surprised this is not available on Node, but indeed it looks like we will need to wait until next year. 
You can write it also just by using reduce as `xs.reduce((acc: B[], val) => acc.concat(f(val)), []);`

File: eslint-bridge/src/rules/x-powered-by.ts [hunk 25]
```diff
@@ -0,0 +1,234 @@
...
+
+function isMethodInvocation(
+  callExpression: estree.CallExpression,
+  objectIdentifierName: string,
+  methodName: string,
+  minArgs: number,
+): boolean {
+  return (
+    callExpression.callee.type === 'MemberExpression' &&
+    callExpression.callee.object.type === 'Identifier' &&
```
Comment (saberduck) [inline eslint-bridge/src/rules/x-powered-by.ts:222]: you can use `isIdentifier` function from `utils`, ie `isIdentifier(callExpression.callee.object, objectIdentifierName)`

File: eslint-bridge/src/rules/x-powered-by.ts [hunk 26]
```diff
@@ -0,0 +1,234 @@
...
+    callExpression.callee.object.name === objectIdentifierName &&
+    callExpression.callee.property.type === 'Identifier' &&
```
Comment (saberduck) [inline eslint-bridge/src/rules/x-powered-by.ts:224]: same as above, use `isIdentifier`

**Review by saberduck — CHANGES_REQUESTED** [review]
looks mostly good, just few minor comments!

File: eslint-bridge/src/rules/x-powered-by.ts [hunk 27]
```diff
@@ -0,0 +1,234 @@
...
+): estree.Node[] {
+  // Invokes `getUniqueWriteUsageOrNode` at most once, from then on
+  // only flattens arrays.
+  function recHelper(nodePossiblyIdentifier: estree.Node): estree.Node[] {
+    return flattenHelper(getUniqueWriteUsageOrNode(context, nodePossiblyIdentifier));
+  }
+
+  // Separate step to ensure progress and avoid `recHelper` invocations without
+  // doing at least one level of unpacking.
+  function flattenHelper(noDerefNode: estree.Node): estree.Node[] {
```
[the remark this answers is not in the document]
Comment (andrey-tyukin-sonarsource) [reply eslint-bridge/src/rules/x-powered-by.ts:128]: It does behave in exactly the same way as the inlined version. The only reason why it exists is that it cannot accidentally recurse on `nodePossiblyIdentifier`, because the variable is not in scope of this helper method. 

Anyway, as it's not clearer, I'll inline it.

File: eslint-bridge/src/rules/x-powered-by.ts (hunk 24 above)
Comment (andrey-tyukin-sonarsource) [reply eslint-bridge/src/rules/x-powered-by.ts:139]: Indeed, it saves couple of lines, but `acc.concat(...)` keeps reallocating a new list every time, whereas `.push(...)` doesn't.

**Review by saberduck — APPROVED** [review]
